# Conditional Statements

## If Statements

In the last mission, we worked with a dataset that stores information for 7,197 mobile apps.

We used lists and for loops to compute the average rating for all of the 7,197 mobile apps. The dataset offers a lot of interesting information, and we might want to answer more granular questions with respect to the average rating:
- What's the average rating of _non-free_ apps?
- What's the average rating of _free_ apps?

We also learned to compute the average value for any list of numbers. However, to answer the two questions above, we first need to find a way to separate free apps from non-free apps because they are all mixed together in our dataset. More specifically, we could:

1. Isolate the ratings for free and non-free apps in separate lists.
2. Compute the average rating for each list.

Before we isolate the ratings for the free apps, let's do a quick recap of how we used the `list_name.append()` command to extract the ratings into a separate list. In the code below, we:

- Start by transforming the `AppleStore.csv` file into a list of lists, and assign that list of lists to a variable named `apps_data`.
- Create an empty list named `ratings`.
- Iterate over `apps_data[1:]` (which excludes the header row), and for each iteration (row), we:
    - Extract the rating and convert it to a float using `float(row[7])`. The rating has the index number `7` and comes as a string, so we need to convert it to a float.
    - We assign the rating to a variable named `rating`.
    - We append `rating` to the `ratings` list we created outside the loop using the `ratings.append(rating)` command.

In [1]:
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()

ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    ratings.append(rating)
    
print(ratings[:5])

[3.5, 4.5, 4.5, 4.5, 4.0]


The problem with our approach above is that it includes all the ratings for both free and non-free apps. To isolate only the ratings of the free apps, we need to add a **condition** to our code above. Specifically, we want to add a rating to the `ratings` list only if the price is equal to `0.0`.

To implement the condition above (If the price is equal to 0.0, then do:) in our code, we can use an `if` **statement**:

In [2]:
ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    price = float(row[4])
    
    if price == 0.0:
        ratings.append(rating)

In the code example above, we iterate over the `apps_data[1:]`. For each iteration, we:

- Assign the rating as a float to a variable named `rating`.
- Assign the price as a float to a variable named `price`. The price also comes as a string, so we need to convert it to a float.
- If the price is equal to `0.0`, we append the rating to the `ratings` list (if the price is `0.0`, then it means the app must be free). Whenever `price` is not equal to `0.0`, the code `ratings.append(rating)` is not executed.

A few things to notice about the `if` statement:

- The `if` statement starts with `if`, it continues with `price == 0.0`, and it ends with `:`.
- We use the `==` operator to check whether price is **equal** to `0.0`. Be careful not to confuse `==` with `=` (recall that `=` is a variable assignment operator in Python, we use it to assign values to variables, and it doesn't tell us anything about equality).
- `ratings.append(rating)` is indented four spaces to the right relative to the `if` statement.

In the next screen, we'll explain `if` statements more in-depth. Until then, let's try to get an intuitive understanding of how `if` statements work by doing an exercise.

---
### Instructions
1. Inside the for loop:
- Assign the price of an app as a float to a variable named `price`. The price is the fifth element in each row (don't forget that the index starts at 0).
- If `price == 0.0`, append the value stored in `rating` to the `free_apps_ratings` list using the `list_name.append()` command (note the `free_apps_ratings` is already defined in the code editor). Be careful with indentation.
2. Outside the for loop body, compute the average rating of free apps. Assign the result to a variable named `avg_rating_free`. The ratings are stored in the `free_apps_ratings` list.

In [3]:
# INITIAL CODE
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()

free_apps_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    # Complete the code from here
    price = float(row[4])
    
    if price == 0.0:
        free_apps_ratings.append(rating)
        
avg_rating_free = sum(free_apps_ratings) / len(free_apps_ratings)
print(avg_rating_free)

3.3767258382642997


---
## Booleans

In the last screen, we used `if price == 0.0` to check whether `price` is equal to `0.0`. When we use the `==` operator to determine whether two values are equal or not, the output returned will always be `True` or `False`:

In [4]:
print(4 == 4)
print(4 == 7)

True
False


In [5]:
price = 0

print(price == 0)
print(price == 2)

True
False


Although they may look like strings, `True` and `False` belong to a different data type (so they are by no means strings):

In [6]:
type(True)

bool

In [7]:
type(False)

bool

`True` and `False` are often called **Boolean values** or **Booleans** — we can see in the code example above that their data type is `bool` ("bool" is an abbreviation for "Boolean").

Boolean values (`True` and `False`) are necessary parts of any `if` statement. `if` must always be followed by one of the following:

1. A Boolean value.
2. An expression that evaluates to a Boolean value.

In [8]:
if True:
    print(100)

100


In [9]:
if 1 == 1:
    print(100)
    
print(1 == 1)

100
True


The code in the **body** of an `if` statement must be indented four spaces to the right (technically, we only need to indent the code at least one space to the right, but the convention in the Python community is to use four spaces per indentation level).

The indented code is _only_ executed when `if` is followed by `True`. When `if` is followed by `False`, the code inside the body is not executed. Notice in the diagram below that `'1 - Output'` and `'3 - Output'` are printed, while `'2 - Output'` is not.

In [10]:
if True:
    print('1 - Output')
    
if False:
    print('2 - Output')
    
if True:
    print('3 - Output')

1 - Output
3 - Output


Note that we can have more than one line of code in the body of an `if` statement. Below, we see three lines of code for each `if` statement.

In [11]:
if True:
    print(1)
    print(2)
    print(3)
    
if False:
    print('A')
    print('B')
    print('C')

1
2
3


---
### Instructions
In the code editor, we've already initialized the variable `a_price` with a value of `0`. Transcribe the following sentences into code by making use of `if` statements:

1. If `a_price` is equal to `0`, then print the string `'This is free'` (remember to use the `==` operator for equality).
2. If `a_price` is equal to `1`, then print the string `'This is not free'`.

In [12]:
a_price = 0

if a_price == 0:
    print('This is free')
if a_price == 1:
    print('This is not free')

This is free


---
## The Average Rating of Non-Free Apps

In the diagram below, we created a list of lists named `app_and_price`, and we want to extract the names of the free apps in a separate list. To do that, we:

- Create an empty list named `free_apps`.
- Iterate over `app_and_price`. For each iteration, we:
    - Extract the name of the app and assign it to a variable named `name`.
    - Extract the price of the app and assign it to a variable named `price`.
    - Append the name of the app to `free_apps` (the empty list that we initialized outside the loop) if the price of the app is equal to `0`.

In [13]:
app_and_price = [['Facebook', 0], ['Instagram', 0], ['Plants vs. Zombies', 0.99], ['Minecraft: Pocket Edition', 6.99], ['Temple Run', 0], ['Plague Inc.', 0.99]]

free_apps = []
for app in app_and_price:
    name = app[0]
    price = app[1]
    
    if price == 0.0:
        free_apps.append(name)
        
print(free_apps)

['Facebook', 'Instagram', 'Temple Run']


The example above should help us better understand what we did in the first screen of this mission, where we managed to extract only the ratings of free apps. At a high level, the steps we took above are identical to the ones we took in the first screen when we used this snippet of code:

In [14]:
ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    price = float(row[4])
    
    if price == 0.0:
        ratings.append(rating)

Above, we:

- Looped through a list of lists named `apps_data`. For every iteration, we:
    - Extracted the rating of the app as a float and assigned it to a variable named `rating`.
    - Extracted the price of the app as a float and assigned it to a variable named `price`.
    - Appended the rating of the app to `ratings` (an empty list that we initialized outside the loop) if the price of the app was equal to `0`.

After we extracted the ratings of the free apps in a separate list, recall that we computed the average value by summing up all the ratings in that list and dividing the sum by the length of that list. However, we still need to compute the average rating for non-free apps.

When we isolated the free apps, we used the condition "if the price **is** equal to 0.0" (`if price == 0.0`). To isolate the non-free apps, we need to change the condition to something like "if the price **is not equal** to 0.0." For "is equal to," we learned that we can use the operator `==`. For "is not equal to," we'll need to use the `!=` operator.

Below, we see an example of how the `!=` operator is used:

In [15]:
print(2 != 0)
print(2 != 2)

True
False


Let's also consider an example where we use a variable (`price`, in the example below) with the `!=` operator:

In [16]:
price = 2

print(price != 0)
print(price != 2)

if price != 0:
    print("Not free")
    
if price != 2:
    print("Price is not equal to 2")

True
False
Not free


In the exercise below, we're going to compute the average rating of non-free apps. In the code editor on the right, we've already added the code we wrote in the first screen of this mission to compute the average rating of free apps.

---
### Instructions

Modify the existing code in the editor on the right to compute the average rating of non-free apps.

1. Change the name of the empty list from `free_apps_ratings` to `non_free_apps_ratings` (the list we defined before the for loop).
2. Change the condition `if price == 0.0` to account for the fact that we now want to isolate only the ratings of non-free apps.
3. Change `free_apps_ratings.append(rating)` to make sure the ratings are appended to the new list `non_free_apps_ratings`.
4. Compute the average value by summing up the values in `non_free_apps_ratings` and dividing by the length of this list. Assign the result to `avg_rating_non_free`.
5. Optional exercise: Inspect the value of `avg_rating_non_free` and compare the average with that of free apps (the average rating of free apps is approximately 3.38 — we computed it in the first screen). Can we use the average values to say that free apps are better than non-free apps, or vice versa?

In [17]:
# INITIAL CODE
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()

non_free_apps_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    price = float(row[4])   
    if price != 0.0:
        non_free_apps_ratings.append(rating)
    
avg_rating_non_free = sum(non_free_apps_ratings) / len(non_free_apps_ratings)

print(avg_rating_non_free)

3.720948742438714


---
## The Average Rating of Gaming Apps

So far, we've used thr `==` and `!=` operators only with integers and floats. But we can use them with other data types as well, such as strings or lists:

In [18]:
print('Games' == 'Music')
print('Games' != 'Music')

False
True


In [19]:
print([1, 2, 3] == [1, 2, 3])
print([1, 2, 3] == [1, 2, 3, 4])

True
False


This enables us to answer more nuanced questions about our dataset, like:

- What's the average rating of gaming apps?
- What's the average rating of non-gaming apps?

Note that the `prime_genre` column describes the app genre, and the genre of gaming apps is encoded as `Games`.

To compute the average rating of gaming apps, we can use the same approach as we took in the previous screen when we computed the average rating of free and non-free apps. In the code example below, we:

- Initialize an empty list named `games_ratings`.
- Loop through `apps_data[1:]`, where `apps_data` is a list of lists that stores our dataset. For each iteration, we:
    - Assign the rating as a float to a variable named `rating`.
    - Assign the genre to a variable named `genre`. The genre will be saved as a string.
    - Append the rating value stored in `rating` to the list `games_rating` if the value in `genre` is equal to the string `'Games'`.
- Compute the average rating of gaming apps, and assign the result to `avg_rating_games`.
- Print `avg_rating_games`.

In [20]:
games_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    
    if genre == 'Games':
        games_ratings.append(rating)
        
avg_rating_games = sum(games_ratings) / len(games_ratings)
print(avg_rating_games)

3.6850077679958573


---
### Instructions

1. Initalize an empty list named `non_games_ratings`.
2. Loop through the `apps_data` list of lists (make sure you don't include the header row. For each iteration of the loop:
    - Assign the rating of the app as a float to a variable named `rating` (the index number of the rating column is `7`).
    - Assign the genre of the app to a variable named `genre` (index number `11`).
    - If the genre is _not_ equal to `'Games'`, append the rating to the `non_games_ratings` list.
3. Compute the average rating of non-gaming apps, and assign the result to a variable named `avg_rating_non_games`.
4. Optional exercise: Compare the average rating of gaming apps (3.69) with that of non-gaming apps. Why do you think we see this difference?

In [21]:
non_games_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    
    if genre != 'Games':
        non_games_ratings.append(rating)
        
avg_rating_non_games = sum(non_games_ratings) / len(non_games_ratings)
print(avg_rating_non_games)

3.343928035982009


---
## Multiple Conditions

So far, we've only worked with single conditions, like:

- If price equals 0.0 (`if price == 0`)
- If genre equals "Games" (`if genre == 'Games'`)

Single conditions won't allow us to answer more granular questions, like:

- What's the average rating of free gaming apps?
- What's the average rating of non-free gaming apps?
- What's the average rating of free non-gaming apps?
- What's the average rating of non-free non-gaming apps?

Fortunately, we can combine two _or more_ conditions together into a single `if` statement using the `and` keyword. As shown below, we can use `and` to simultaneously check whether an app is both free and has a gaming genre.

In [22]:
app1_price = 0
app1_genre = 'Games'

if app1_price == 0 and app1_genre == 'Games':
    print('This is a free game!')
    
print(app1_price == 0 and app1_genre == 'Games')

This is a free game!
True


In [23]:
app2_price = 19
app2_genre = 'Games'

if app2_price == 0 and app2_genre == 'Games':
    print('This is a free game!')
    
print(app2_price == 0 and app2_genre == 'Games')

False


Notice above that code like `app1_price == 0 and app1_genre == 'Games'` outputs a single Boolean value.

Python evaluates any combination of Booleans into a single Boolean value:

In [24]:
print(True and True)
print(True and False)
print(False and True)
print(False and False)

True
False
False
False


As a general rule, when we combine Booleans using `and`, the resulting Boolean is `True` only if all the Booleans are `True`. If any of the Booleans are `False`, then the resulting Boolean will be `False`:

In [25]:
print(True and True and True)
print(True and True and False)
print(True and True and True and False)
print(True and True and True and True)

True
False
False
True


---
### Instructions

Complete the code in the editor to compute the average rating of free gaming apps.

1. Inside the for loop, append the rating to the `free_games_rating` list if the price is equal to `0.0` **and** the genre is equal to `'Games'`.
2. Outside the for loop, compute the average rating of free gaming apps. Assign the result to a variable named `avg_rating_free_games`.

In [26]:
# INITIAL CODE
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()

free_games_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    price = float(row[4])
    genre = row[11]
    # Complete code from here
    if price == 0.0 and genre == 'Games':
        free_games_ratings.append(rating)
    
avg_rating_free_games = sum(free_games_ratings) / len(free_games_ratings)
print(avg_rating_free_games)

3.5285777580859548


---
## The `or` Operator

If we look at the first five apps, we can see in the `prime_genre` column that Facebook's genre is "Social Networking" while CLash of Clans' and Temple Run's is "Games".

Social networking apps and games are usually popular and addictive, and we may want to further investigate this category. One thing we might want to find out is the average rating of this category that encompasses both games and social networking apps.

To do that, we first need to isolate the ratings of all the apps whose genre is either "Social Networking" _or_ "Games" into a separate list. Then, we can compute the average value using techniques we already know.

If we wanted to isolate the ratings of these apps using the condition `if genre == 'Social Networking' and genre == 'Games'`, we'd end up with an empty list because there's no app whose genre is both "Social Networking" and Games" -- an app can only have a single genre.

In [27]:
games_social_ratings = []

for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    
    if genre == 'Social Networking' and genre == 'Games':
        games_social_ratings.append(rating)
        
print(games_social_ratings)
len(games_social_ratings)

[]


0

We need to isolate the rating of an app only if the genre is "Social Networking" **or** "Games", not "Social Networking" **and** "Games". To account for this difference, we can use `or` instead of `and`:

In [28]:
games_social_ratings = []

for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    
    if genre == 'Social Networking' or genre == 'Games':
        games_social_ratings.append(rating)
        
print(games_social_ratings[:5])
len(games_social_ratings)

[3.5, 4.5, 4.5, 4.5, 4.5]


4029

We call `or` and `and` **logical operators**. Just like `and`, `or` bridges two or more Booleans together. As we learned previously, Python evaluates any combination of Booleans to a single Boolean value:

In [29]:
print(True or True)
print(True or False)
print(False or False)
print(False or True)

True
True
False
True


In [30]:
print(True and True)
print(True and False)
print(False and True)
print(False and False)

True
False
False
False


When we combine Booleans using `or`, the resulting Boolean is `False` only if all the Booleans are `False`. If any of the Booleans are `True`, then the resulting is `True`:

In [31]:
print(False or False or False)
print(False or False or False or True)

False
True


Returning to our apps example, the condition `if genre == 'Social Networking' or genre == 'Games'` will only resolve to `False` when an app's genre is neither "Social Networking" nor "Games". Otherwise, it will resolve to `True`.

Now let's practice using the `or` operator by compuyting the average rating og the apps whose genre is either "Social Networking" or "Games".

---
### Instructions

Complete the code in the editor to compute the average rating of the apps whose genre is either "Social Networking" or "Games".

1. Inside the for loop, append the rating to the `games_social_ratings` list if the genre is either `'Social Networking'` **or** `'Games'`.
2. Outside the for loop, compute the average rating of the apps whose genre is either "Social Networking" or "Games", and assign the result to a variable named `avg_games_social`.

In [32]:
# INITIAL CODE
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()

games_social_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    # Complete code from here
    if genre == 'Social Networking' or genre == 'Games':
        games_social_ratings.append(rating)
        
avg_games_social = sum(games_social_ratings) / len(games_social_ratings)
print(avg_games_social)

3.655994043186895


---
## Combining Logical Operators

In the previous exercise, we computed the average rating of the apps whose genre is either "Social Networking" or "Games". We can ask even more specific questions, like:

- What is the average rating of _free_ apps whose genre is either "Social Networking" or "Games"?
- What is the average rating of _non-free_ apps whose genre is either "Social Networking" or "Games"?

To answer the first question, we need to isolate the apps that:

- Are in either the "Social Networking" **or** "Games" genre
- **And** have a price of `0.0`

To isolate these apps, we can combine `or` with `and` in a single `if` statement:

In [34]:
free_games_social_ratings = []

for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    price = float(row[4])
    
    if (genre == "Social Networking" or genre == "Games") and price == 0.0:
        free_games_social_ratings.append(rating)

Notice that we enclosed the `genre == 'Social Networking' or genre == 'Games` part within parentheses. This helps Python understand the specific logic we want for our `if` statement.

If we don't use the parentheses to make our logic explicity, Python defaults will lead us to unwanted results -- like incorrectly including non-free apps and can cause **logical errors**.

In the case of a syntax or runtime error, the computer stops the code from running and raises an error. For logical errors, the computer doesn't raise any error -- the code runs just fine and the logical error slips through. To prevent this, we should try to pay extra attention to the logic of our code.

---
### Instructions

1. Compute the average rating of non-free apps whose genre is either "Social Networking" or "Games".
- Assign the result to a variable named `avg_non_free`.

In [35]:
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()

free_games_social_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    price = float(row[4])
    
    if (genre == 'Social Networking' or genre == 'Games') and price == 0:
        free_games_social_ratings.append(rating)
        
avg_free = sum(free_games_social_ratings) / len(free_games_social_ratings)

# Non-free apps (average)
nonfree_games_social_ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    price = float(row[4])
    
    if (genre == 'Social Networking' or genre == 'Games') and price != 0.0:
        nonfree_games_social_ratings.append(rating)
        
avg_non_free = sum(nonfree_games_social_ratings) / len(nonfree_games_social_ratings)

print(avg_free)
print(avg_non_free)

3.496875
3.8904235727440146


---
## The `not` Operator

Earlier we saw that we can use `!=` to check if two objects are different, in other words, to check if they are not equal.

But what if we want to check if a more complex condition is **not** verified?

In the example on the previous screen, we computed the average rating of free apps whose genre is either "Social Networking" or "Games". But what if we wanted to compute the average rating of free apps whose genre does not verify the condition `genre == "Social Networking" or genre == "Games"`? The `not` operator can help us with that.

In [36]:
free_not_games_social_ratings = []

for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    price = float(row[4])
    
    if not (genre == "Social Networking" or genre == "Games") and price == 0.0:
        free_not_games_social_ratings.append(rating)

We just have to add the keyword `not` to the left of the condition we wish to "negate".

This operator is different from `and` and `or` in a fundamental way. There as `and` and `or` require two conditions to be evaluated, `not` requires only one. Because of this we say that `and` and `or` are **binary operators**, and `not` is an **unary operator**.

Just like `and` and `or`, `not` returns a Boolean value. More precisely, what it does is change `True` to `False` and `False` to `True`:

In [37]:
print(not True)
print(not False)

False
True


---
### Instructions

1. Compute the average rating of non-free apps whose genres do not verify the condition `genre == 'Social Networking' or genre == 'Games'`. Assign them to `avg_non_free_non_sn_games`.

In [38]:
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()

non_free_non_sn_games = []
for row in apps_data[1:]:
    rating = float(row[7])
    genre = row[11]
    price = float(row[4])
    
    # Complete code here
    if not (genre == "Social Networking" or genre == "Games") and price != 0.0:
        non_free_non_sn_games.append(rating)
        
avg_non_free_non_sn_games = sum(non_free_non_sn_games) / len(non_free_non_sn_games)

print(avg_non_free_non_sn_games)

3.5383597883597884


---
## Comparison Operators

Previously, we used the `==` and `!=` operators to check whether two values are equal or not. When we check for equality, we _compare_ one value to another to be able to determine whether they are equal or not. For this reason, we call `==` and `!=` **comparison operators**.

We can compare value `A` to value `B` to determine whether:

- `A` is **equal** to `B` and vice versa (`B` is equal to `A`).
- `A` is **not equal** to `B` and vice versa.
- `A` is **greater** than `B` and vice versa.
- `A` is **greater than or equal to** `B` or vice versa.
- `A` is **less** than `B` or vice versa.
- `A` is **less than or equal to** `B` or vice versa.

In Python, we have special operators for each of the comparison operations.

Just like with `==` and `!=`, comparing values using any of the comparison operators above will output a single Boolean value:

In [39]:
print(8 > 1)
print(8 < 1)
print(30 >= 30)
print(20 <= 19)

True
False
True
False


The fact that a comparison outputs a single Boolean value allows us to use a comparison inside an `if` statement:

In [40]:
app_name = "Ulyssses"
app_price = 24.99

if app_price > 20:
    print('This app is expensive!')
    
print(app_price > 20)

This app is expensive!
True


These new comparison operators allow us to answer more granular questions about our dataset:

- How many apps have a reating of 4.0 or greater?
- What is the average rating of the apps that have a price greater than \$9?
- How many apps have a price greater than \$9?
- How many apps have a price smaller than or equal to \$9?

Using what we know, we can answer the first question in at least two ways.

In [41]:
apps_4_or_greater = []

for row in apps_data[1:]:
    rating = float(row[7])
    if rating >= 4.0:
        apps_4_or_greater.append(rating)
        
len(apps_4_or_greater)

4781

Above, we:

- Initialized an empty list named `apps_4_or_greater`.
- Looped through `apps_data[1:]`, and for every iteration, we:
    - Stored the rating value as a float to a variable named `rating`.
    - Appended the rating to `apps_4_or_greater` if the value stored in `rating` was greater than or equal to `4.0`.
- Measured the length of the `apps_4_or_greater` list to find out the number of apps that have a rating of `4.0` or greater.
    - After the loop, the `apps_4_or_greater` list will only store the ratings of the apps that were rated `4.0` or better. The list `apps_4_or_greater` has 4,781 ratings that are `4.0` or greater, which means that there are 4,781 apps that have a rating of `4.0` or greater.
    
We can also use another approach to answer the question, where we initialize a variable with a value of `0` and then increment that variable by `1` each time we find a rating of `4.0` or greater:

In [42]:
n_of_apps = 0

for row in apps_data[1:]:
    rating = float(row[7])
    if rating >= 4.0:
        n_of_apps = n_of_apps + 1
        
print(n_of_apps)

4781


Above, we:

- Initialized a variable of `n_of_apps` with a value of `0`
- Looped through `apps_data[1:]`, and for every iteration, weL
    - Stored the rating value as a float to a variable named `rating`
    - Incremented the value of `n_of_apps` by `1` if the value stored in `rating` was greater than or equal to `4.0`
- Printed the value of `n_of_apps` to find out the number of apps that have a rating of `4.0` or greater
    - `n_of_apps` was incremented by `1` during the loop every time a rating was `4.0` or greater. The value of `n_of_apps` is 4,781 after the looping, which means that 4,781 ratings are `4.0` or greater. Each rating belongs to an app, so this means that there are 4,781 apps that have a rating of `4.0` or greater.
    
It's worth noting that comparison between different data types isn't straightfoward. For instance, even though it is not true that the integer `1` is greater than the character `a` in an obvious way, `print(1 > 'a')` will not print `False`:

In [43]:
print(1 > 'a')

TypeError: '>' not supported between instances of 'int' and 'str'

We get an error because in Python this comparison is not meaningful.

Now let's answer the other three questions we asked above:
- What is the average rating of the apps that have a price greater than \$9?
- How many apps have a price greater than \$9?
- How many apps have a price smaller than or equal to \$9?

---
### Instructions

1. Compute the average rating of the apps that have a price greater than \$9.
    - Using a for loop, isolate the ratings of all the apps that have a price greater than \$9. When you iterate over `apps_data`, make sure you don't include the header now.
    - Find the average value of these ratings and assign the result to a variable named `avg_rating`.
2. Find out how many apps have a price greater than \$9 and assign the result to a variable named `n_apps_more_9`. You can use the list of ratings from the previous question to find the answer.
3. Find out how many apps have a price less than _or equal to_ \$9 and assign the result to a variable named `n_apps_less_9`. The list of ratings from the first question can help you find a quick answer.

In [44]:
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()
ratings = []
for row in apps_data[1:]:
    rating = float(row[7])
    price = float(row[4])
    
    if price > 9:
        ratings.append(rating)
        
avg_rating = sum(ratings) / len(ratings)
n_apps_more_9 = len(ratings)
n_apps_less_9 = len(apps_data[1:]) - len(ratings)

---
## The `else` Clause

Let's say we need to use information from the `price` column to label each app as "free" or "non-free". If the price is equal to `0.0`, we want to label the app "free". Otherwise, we want to label it "non-free".

Remember that we store our dataset as a list of lists -- each row is represented as a list that describes an app. To label the app, we want to add the string `'free'` or `'non-free'` at the end of each list (row). On a smaller scale, this is what we want to do (below, we're just using a small extract from our dataset, showing just the name and the price of four apps):

In [45]:
apps_data = [['Call of Duty: Zombies', 5.0], ['Facebook', 0.0], ['Instagram', 0.0], ['Temple Run', 0.0]]

for app in apps_data:
    price = app[1]
    
    if price == 0.0:
        app.append('free')
    if price != 0.0:
        app.append('non-free')
        
print(apps_data)

[['Call of Duty: Zombies', 5.0, 'non-free'], ['Facebook', 0.0, 'free'], ['Instagram', 0.0, 'free'], ['Temple Run', 0.0, 'free']]


In the code above, we iterated through the `apps_data` list of lists, and for each iteration:
- We saved the price value to a variable named `price`.
- If `price == 0.0`, we appended the string `'free'` to the list `app` (`app` is the iteration variable).
- If `price != 0.0`, we appended the string `'non-free'` to the list `app`.

For each iteration, the computer checked whether `price == 0.0` and `price != 0.0`. But once we know that `price == 0.0` for an app, it's redundant to also check `price != 0.0` for the same app -- if we know that the price is `0`, it doesn't make logical sense to also check whether the price is different than `0`.
    
In our small dataset above, we have three free apps. The computer evaluated `price == 0.0` as `True` three times, and then checked whether `price != 0.0` for the same number of times. THis means that the computer performed three redundant operations -- it still evaluated `price != 0.0` while knowing that `price == 0.0` is `True`.

If we have a dataset with 5,000 free apps, this would mean 5,000 redundant operations. We can avoid this redundancy by combining an `if` statement with an `else` **clause**:

In [46]:
apps_data = [['Call of Duty: Zombies', 5.0], ['Facebook', 0.0], ['Instagram', 0.0], ['Temple Run', 0.0]]

for app in apps_data:
    price = app[1]
    
    if price == 0.0:
        app.append('free')
    else:
        app.append('non-free')
        
print(apps_data)

[['Call of Duty: Zombies', 5.0, 'non-free'], ['Facebook', 0.0, 'free'], ['Instagram', 0.0, 'free'], ['Temple Run', 0.0, 'free']]


The code within the body of an `else` clause is executed _only_ if the `if` statement that precedes it resolves to `False`.

In [47]:
if False:
    print(1)
else:
    print(2)

2


In [48]:
if True:
    print(1)
else:
    print(2)

1


Consider also an example that's similar to our apps example above:

In [49]:
price = 5

if price == 0:
    print('free')
else:
    print('not free')

not free


In [50]:
price = 0

if price == 0:
    print('free')
else:
    print('not free')

free


In our apps example above, the code within the body of the `else` clause is executed only if `price == 0.0` evaluates to `False`. If `price == 0.0` is `True`, then the code `app.append('free')` is executed, and the computer moves forward _without_ executing the `else` clause.

Note that an `else` clause must be combined with a preceding `if` statement. We can have an `if` statement without an `else` clause, but we can't have an `else` clause without a preceding `if` statement.

In [51]:
else:
    print(1)

SyntaxError: invalid syntax (<ipython-input-51-5f6ba2ad6b3b>, line 1)

When we combine a statement with a clause, we create a **compound statement** -- combining an `if` statement with an `else` clause makes up a compound statement.

---
### Instructions

1. Complete the code in the editor to label each app as "free" or "non-free" depending on its price.
- Inside the for loop:
    - **If** the price of the app is `0.0`, then label the app as "free" by appending the string `'free'` to the current iteration variable.
    - **Else**, label the app "non-free" by appending the string `'non-free'` to the current iteration variable. Make sure you don't write `'non_free'` instead of `'non-free'`.
- By adding labels to the end of each row, we basically created a new column. Name this column "free_or_not" by appending the string `'free_or_not'` to the first row of the `apps_data` dataset. Make sure this is done outside the for loop.
2. Print the header row and the first five rows to see some of the changes we made.

In [52]:
# INITIAL CODE
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()


for app in apps_data[1:]:
    price = float(app[4])
    # Complete code from here
    if price == 0.0:
        app.append('free')
    else:
        app.append('non-free')
        
apps_data[0].append('free_or_not')
print(apps_data[:6])

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic', 'free_or_not'], ['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1', 'free'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1', 'free'], ['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1', 'free'], ['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1', 'free'], ['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1', 'free']]


---
## The `elif` Clause

Let's say we need to do a more granular labeling rather than just using "free" and "non-free". We want to label the app using this convention:

| **price** | **label** |
| :--- | :--- |
| 0 | free |
| < 20 | affordable |
| 20 - 50 | expensive |
| > 50 | very expensive |

Using what we know, we can only do the transformations above using a combination of `if` statements.

In [54]:
apps_data = [['Facebook', 0.0],
            ['Notion', 14.99],
            ['Astropad Standard', 29.99],
            ['NAVIGON Europe', 74.99]]

for app in apps_data:
    price = app[1]
    
    if price == 0.0:
        app.append('free')
    if price > 0.0 and price < 20:
        app.append('affordable')
    if price >= 20 and price < 50:
        app.append('expensive')
    if price >= 50:
        app.append('very expensive')
        
print(apps_data)

[['Facebook', 0.0, 'free'], ['Notion', 14.99, 'affordable'], ['Astropad Standard', 29.99, 'expensive'], ['NAVIGON Europe', 74.99, 'very expensive']]


When an app is free, `price == 0.0` evaluates to `True` and `app.append('free')` is executed. But then the computer continues to do redundant operations -- it checks whether:

- `price > 0 and price < 20`
- `price >= 20 and price < 50`
- `price > 50`

We already know the three conditions above will evaluate to `False` once we find out that `price == 0.0` is `True`. To stop the computer from doing redundant operations, we can use `elif` **clauses**:

In [55]:
apps_data = [['Facebook', 0.0],
            ['Notion', 14.99],
            ['Astropad Standard', 29.99],
            ['NAVIGON Europe', 74.99]]

for app in apps_data:
    price = app[1]
    
    if price == 0.0:
        app.append('free')
    elif price > 0.0 and price < 20:
        app.append('affordable')
    elif price >= 20 and price < 50:
        app.append('expensive')
    elif price >= 50:
        app.append('very expensive')
        
print(apps_data)

[['Facebook', 0.0, 'free'], ['Notion', 14.99, 'affordable'], ['Astropad Standard', 29.99, 'expensive'], ['NAVIGON Europe', 74.99, 'very expensive']]


The code within the body of an `elif` clause is executed _only_ if:

- The preceding `if` statement (or other preceding `elif` clauses) resolves to `False`; **and**
- The condition specified after the `elif` keyword evaluates to `True`.

In the case above, if `price == 0.0` is `True`, the computer executes `app.append('free')` and moves forward _without_ executing any of the following `elif` clauses.

If `price > 0.0 and price < 20` is `True`, then `app.append('affordable')` is executed, and the remaining two `elif` clauses are not considered anymore. If `price >= 20 and price < 50` is `True`, then `app.append('expensive')` is executed, and the last `elif` clause is not considered anymore.

It's possible to replace the last `elif` clause with an `else`clause. For our data above, the output would be identical.

Note, however, that the logic is different when we use an `else` clause. For `elif price >= 50`, `app.append('very expensive')` will be executed only if the price is greater than or equal to `50`. For `else`, there's no condition to be met (other than the previous `if` and `elif` clauses resolving to `False`), and `app.append('very expensive')` will be executed even if the price has a value of `-5` or `-100`.

Just like `else`, the `elif` clause must be combined with a preceding `if` statement and can't stand on its own:

In [56]:
elif:
    print(1)

SyntaxError: invalid syntax (<ipython-input-56-94e5a6a9bcb8>, line 1)

---
### Instructions

1. Complete the code in the editor to label each app as "free", "affordable", "expensive", or "very expensive". Inside the loop:
- **If the price of the app is `0`**, label the app as "free" by appending the string `'free'` to the curent iteration variable.
- **If the price of the app is greater than `0` and less than `20`**, label the app as "affordable". For efficiency purposes, use an `elif` clause.
- **If the app is greater than or equal to `20` and less than `50`**, label the app as "expensive". For efficiency purposes, use an `elif` clause.
- **If the app is greater than or equal to `50`**, label the app as "very expensive". For efficiency purposes, use an `elif` clause.

2. Name the newly created column "price_label" by appending the string `'price_label'` to the first row of the `apps_data` dataset.
3. Inspect the header row and the first five rows to see some of the changes you made.

In [57]:
# INITIAL CODE
opened_file = open('AppleStore.csv')
from csv import reader
read_file = reader(opened_file)
apps_data = list(read_file)
opened_file.close()

for app in apps_data[1:]:
    price = float(app[4])
    # Complete code from here
    if price == 0.0:
        app.append('free')
    elif price >= 0 and price < 20:
        app.append('affordable')
    elif price >= 20 and price < 50:
        app.append('expensive')
    elif price > 50:
        app.append('very expensive')
        
apps_data[0].append('price_label')
print(apps_data[:6])

[['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic', 'price_label'], ['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1', 'free'], ['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1', 'free'], ['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1', 'free'], ['420009108', 'Temple Run', '65921024', 'USD', '0.0', '1724546', '3842', '4.5', '4.0', '1.6.2', '9+', 'Games', '40', '5', '1', '1', 'free'], ['284035177', 'Pandora - Music & Radio', '130242560', 'USD', '0.0', '1126879', '3594', '4.0', '4.5', '8.4.1', '12+', 'Music', '37', '4', '1', '1', 'free']]
